### PR TITLE
Integrate BrightIdRegister with JurorsRegistry and add max activated stake per juror restriction.

### PR DESCRIPTION
**Pull request by willjgriff**
Converted to Buidler to enable compilation of different Solidity versions using 2 buidler configs. Specifically the BrightIdRegister uses v0.4.24 (as it's an AragonApp) and Celeste uses v0.5.8. 

The court is structured with a global config that is accessible in each of the individual modules eg JurorsRegistry, Subscriptions, DisputeManager etc. 

I updated the courts global config to include `minMaxPctTotalSupply` and `maxMaxPctTotalSupply` which are used to determine the max a juror can stake. Specifically they are the min and max % of the total supply of the jurors token which will be used as a jurors max stake depending on how much is currently staked. When there is 0 staked, the max juror stake is maxMaxPctTotalSupply percent of the total supply. When all of the jurors tokens are staked, the max juror stake is minMaxPctTotalSupply percent of the total supply.

I added the BrightIdRegister to the global config as a new module.

I Integrated the BrightIdRegister to the `JurorsRegistry` so we can check when a juror activates that they are not activating more than the max stake. If the calling address is not registered with the BrightIdRegister, it will revert. Note jurors using new accounts will have to register with the BrightIdRegister before being able to activate.

### CODE REVIEW

File: contracts/registry/JurorsRegistry.sol [hunk 1]
```diff
@@ -498,6 +510,37 @@ contract JurorsRegistry is ControlledRecoverable, IJurorsRegistry, ERC900, Appro
...
+    * @param _juror Any of the BrightId accounts registered addresses
+    * @return Amount of active tokens of a juror that are not locked due to ongoing disputes
+    */
+    function jurorsTotalActiveStake(address _juror) external view returns (uint256) {
+        address jurorBrightId = _jurorBrightId(_juror);
+        return brightIdActiveStake[jurorBrightId];
+    }
+
+    /**
+    * @dev Tell the max balance a juror can currently activate, calculated using a sliding scale of a percent of the
```
Comment (fabriziovigevani) [inline contracts/registry/JurorsRegistry.sol]:
```suggestion
    * @dev Tell the max balance a juror can currently activate, calculated using a sliding scale of a percent of the total supply
```

File: contracts/registry/JurorsRegistry.sol [hunk 2]
```diff
@@ -498,6 +510,37 @@ contract JurorsRegistry is ControlledRecoverable, IJurorsRegistry, ERC900, Appro
...
+            return 0;
+        }
+
+        // TODO: Can this calculation be minimised?
+        uint256 minMaxPctTotalSupply = _getConfigAt(_termId).jurors.minMaxPctTotalSupply;
+        uint256 maxMaxPctTotalSupply = _getConfigAt(_termId).jurors.maxMaxPctTotalSupply;
+        uint256 diffOfPct = maxMaxPctTotalSupply - minMaxPctTotalSupply; // No need for safemath, we ensure min is less than max in config setting
+        uint256 totalActiveBalanceAtTerm = _totalActiveBalanceAt(_termId);
+
+        uint256 currentPctOfTotalSupply = maxMaxPctTotalSupply.sub(totalActiveBalanceAtTerm.mul(diffOfPct).div(jurorsTokenTotalSupply));
```
Comment (fabriziovigevani) [inline contracts/registry/JurorsRegistry.sol]: Could we use normal division, as we are early returning if `jurorsTokenTotalSupply === 0`?
```suggestion
        uint256 currentPctOfTotalSupply = maxMaxPctTotalSupply.sub(totalActiveBalanceAtTerm.mul(diffOfPct) / jurorsTokenTotalSupply);
```

File: contracts/registry/JurorsRegistry.sol [hunk 3]
```diff
@@ -565,16 +608,35 @@ contract JurorsRegistry is ControlledRecoverable, IJurorsRegistry, ERC900, Appro
...
+            uint256 newActiveBalance = activeBalance.add(amountToActivate);
+            require(newActiveBalance >= minActiveBalance, ERROR_ACTIVE_BALANCE_BELOW_MIN);
+
+            // Due to min active balance living at the global config level, we can't ensure it is less than the max
+            // active balance when set, because max active balance is determined using the juror token, specific to this
+            // contract. Therefore we revert if the min active balance is more than the max active balance.
+            require(minActiveBalance < maxActiveBalance, ERROR_ACTIVE_BALANCE_LIMITS_INCONSISTENT);
+
+            uint256 jurorsNewTotalActiveStake = _updateBrightIdActiveStake(_juror, amountToActivate, true);
+            require(jurorsNewTotalActiveStake <= maxActiveBalance, ERROR_ACTIVE_BALANCE_ABOVE_MAX);
```
Comment (fabriziovigevani) [inline contracts/registry/JurorsRegistry.sol]: Might make sense to move this outside the `if else` block as it's not dependent on the jurors existence.

File: contracts/registry/JurorsRegistry.sol [hunk 4]
```diff
@@ -730,6 +795,7 @@ contract JurorsRegistry is ControlledRecoverable, IJurorsRegistry, ERC900, Appro
     */
     function _setTotalActiveBalanceLimit(uint256 _totalActiveBalanceLimit) internal {
         require(_totalActiveBalanceLimit > 0, ERROR_BAD_TOTAL_ACTIVE_BALANCE_LIMIT);
+        require(_totalActiveBalanceLimit > maxActiveBalance(_getLastEnsuredTermId()), ERROR_TOTAL_ACTIVE_BALANCE_TOO_HIGH);
```
Comment (fabriziovigevani) [inline contracts/registry/JurorsRegistry.sol]: Should it be too low?
```suggestion
        require(_totalActiveBalanceLimit > maxActiveBalance(_getLastEnsuredTermId()), ERROR_TOTAL_ACTIVE_BALANCE_TOO_LOW);
```

File: contracts/court/config/CourtConfig.sol [hunk 5]
```diff
@@ -21,6 +21,10 @@ contract CourtConfig is IConfig, CourtConfigData {
     string private constant ERROR_BAD_APPEAL_STEP_FACTOR = "CONF_BAD_APPEAL_STEP_FACTOR";
     string private constant ERROR_ZERO_COLLATERAL_FACTOR = "CONF_ZERO_COLLATERAL_FACTOR";
     string private constant ERROR_ZERO_MIN_ACTIVE_BALANCE = "CONF_ZERO_MIN_ACTIVE_BALANCE";
+    string private constant ERROR_MIN_MAX_TOTAL_SUPPLY_ZERO = "CONF_MIN_MAX_TOTAL_SUPPLY_ZERO";
+    string private constant ERROR_INVALID_MAX_MAX_TOTAL_SUPPLY_PCT = "CONF_INVALID_MAX_MAX_TOTAL_SUPPLY_PCT";
+    string private constant ERROR_MIN_MORE_THAN_MAX_ACTIVE_PCT = "CONF_MIN_MORE_THAN_MAX_ACTIVE_PCT";
+    string private constant ERROR_JURORS_FOR_MIN_ACTIVE_ZERO = "CONF_JURORS_FOR_MIN_ACTIVE_ZERO";
```
Comment (fabriziovigevani) [inline contracts/court/config/CourtConfig.sol]: What's this error for?

File: test/court/controller/controller-config.js [hunk 6]
```diff
@@ -131,6 +135,18 @@ contract('Controller', ([_, configGovernor, someone, drafter, appealMaker, appea
       it('cannot use a min active balance 0', async () => {
         await assertRevert(courtHelper.deploy({ minActiveBalance: bn(0) }), CONFIG_ERRORS.ZERO_MIN_ACTIVE_BALANCE)
       })
+
+      it('cannot use a min max pct total supply active balance 0', async () => {
+        await assertRevert(courtHelper.deploy({ minMaxPctTotalSupply: bn(0)}), "CONF_MIN_MAX_TOTAL_SUPPLY_ZERO")
```
Comment (fabriziovigevani) [inline test/court/controller/controller-config.js:140]: Consider adding all error strings to `CONFIG_ERRORS`

File: contracts/registry/JurorsRegistry.sol [hunk 7]
```diff
@@ -498,6 +510,37 @@ contract JurorsRegistry is ControlledRecoverable, IJurorsRegistry, ERC900, Appro
         return _currentUnlockedActiveBalanceOf(juror);
     }
 
+    /**
+    * @dev Tell the total amount of active tokens for the jurors BrightId account
+    * @param _juror Any of the BrightId accounts registered addresses
+    * @return Amount of active tokens of a juror that are not locked due to ongoing disputes
```
Comment (fabriziovigevani) [inline contracts/registry/JurorsRegistry.sol]: If we are considering active stake which is not locked, should we do a 
```
 _updateBrightIdActiveStake(jurorAddress, draftLockAmount, false);
```
 when a juror is drafted?

Comment (fabriziovigevani) [reply contracts/registry/JurorsRegistry.sol]: Ah actually i see that we reduce the locked amount only when slashing.
So we should remove the `that are not locked` part?

File: test/registry/jurors-registry-activation.js [hunk 8]
```diff
@@ -1,35 +1,84 @@
...
+  const PCT_BASE_HIGH_PRECISION = bigExp(1, 18) // 100%
 
   const MIN_ACTIVE_AMOUNT = bigExp(100, 18)
+  const MIN_MAX_PCT_TOTAL_SUPPLY = bigExp(1, 15) // 0.1%
+  const MAX_MAX_PCT_TOTAL_SUPPLY = bigExp(1, 16) // 1%
   const TOTAL_ACTIVE_BALANCE_LIMIT = bigExp(100e6, 18)
 
-  before('create base contracts', async () => {
-    controller = await buildHelper().deploy({ minActiveBalance: MIN_ACTIVE_AMOUNT })
+  const maxActiveBalanceForConfig = async (minMaxPctTotalSupply, maxMacPctTotalSupply, totalActiveBalance) => {
```
Comment (fabriziovigevani) [inline test/registry/jurors-registry-activation.js]:
```suggestion
  const maxActiveBalanceForConfig = async (minMaxPctTotalSupply, maxMaxPctTotalSupply, totalActiveBalance) => {
```

File: test/registry/jurors-registry-activation.js [hunk 9]
```diff
@@ -1,35 +1,84 @@
...
+    const diffOfPct = maxMacPctTotalSupply.sub(minMaxPctTotalSupply)
+    const anjTotalSupply = await ANJ.totalSupply()
+    const currentPctOfTotalSupply = maxMacPctTotalSupply.sub(diffOfPct.mul(totalActiveBalance).div(anjTotalSupply))
+    return (await ANJ.totalSupply()).mul(currentPctOfTotalSupply).div(PCT_BASE_HIGH_PRECISION)
+  }
+
+  const maxActiveBalanceAtTermForConfig = async (termId, minMaxPctTotalSupply, maxMacPctTotalSupply) => {
```
Comment (fabriziovigevani) [inline test/registry/jurors-registry-activation.js]:
```suggestion
  const maxActiveBalanceAtTermForConfig = async (termId, minMaxPctTotalSupply, maxMaxPctTotalSupply) => {
```

File: test/registry/jurors-registry-activation.js [hunk 10]
```diff
@@ -82,45 +135,98 @@ contract('JurorsRegistry', ([_, juror]) => {
...
+          assertBn(activeAddressCurrentTotalActiveStake, activeAddressPreviousTotalActiveStake.add(activationAmount), 'incorrect total active stake for active account')
+          assertBn(brightIdAddressCurrentTotalActiveStake, brightIdAddressPreviousTotalActiveStake.add(activationAmount), 'incorrect total active stake for brightId account')
+        })
+
+        if (!requestedAmount.eq(bn(0)) && !requestedAmount.eq(USE_MAX_ACTIVE_AMOUNT_FOR_0_JURORS)) {
+          it('updates the total active balance correctly when a different juror with the same brightid is used', async () => {
+            requestedAmount = await useAmount(requestedAmount)
+            await ANJ.generateTokens(jurorBrightIdAddress, MIN_ACTIVE_AMOUNT)
+            await ANJ.approveAndCall(registry.address, MIN_ACTIVE_AMOUNT, '0x', { from: jurorBrightIdAddress })
+            const { active: previousActiveBalance, available: previousAvailableBalance, locked: previousLockedBalance, pendingDeactivation: previousDeactivationBalance } = await registry.balanceOf(jurorActiveAddress)
```
Comment (fabriziovigevani) [inline test/registry/jurors-registry-activation.js]: Seem to not be used.

File: test/registry/jurors-registry-activation.js [hunk 11]
```diff
@@ -134,34 +240,40 @@ contract('JurorsRegistry', ([_, juror]) => {
         })
 
         it('emits an activation event', async () => {
+          requestedAmount = await useAmount(requestedAmount)
+          deactivationAmount = await useAmount(deactivationAmount)
```
Comment (fabriziovigevani) [inline test/registry/jurors-registry-activation.js]: Might not be necessary as we don't use it in this test.

File: contracts/registry/JurorsRegistry.sol (hunk 4 above)
Comment (fabriziovigevani) [reply contracts/registry/JurorsRegistry.sol]: Also, i think we missed a test for this error.

File: test/registry/jurors-registry-activation.js [hunk 12]
```diff
@@ -705,4 +907,75 @@ contract('JurorsRegistry', ([_, juror]) => {
...
+
+      let termId
+
+      beforeEach(async () => {
+        await ANJ.generateTokens(jurorActiveAddress, TOTAL_ACTIVE_BALANCE_LIMIT)
+        termId = await controller.getLastEnsuredTermId()
+      })
+
+      it('returns correct value', async () => {
+        assertBn(await registry.maxActiveBalance(await controller.getLastEnsuredTermId()), await currentMaxActiveBalance(), 'Incorrect max active balance')
```
Comment (fabriziovigevani) [inline test/registry/jurors-registry-activation.js]:
```suggestion
        assertBn(await registry.maxActiveBalance(termId, await currentMaxActiveBalance(), 'Incorrect max active balance')
```

**Review by fabriziovigevani — COMMENTED** [review]
Quite the PR! Excellent job man! 

I've yet to see the slashing and staking tests, but some comments for now.

File: test/court/controller/controller-config.js (hunk 6 above)
Comment (willjgriff) [reply test/court/controller/controller-config.js:140]: I'd rather not bother with that. I think adding them to that file only makes it longer to determine what they are.

File: test/registry/jurors-registry-slashing.js [hunk 13]
```diff
@@ -162,19 +175,47 @@ contract('JurorsRegistry', ([_, juror, secondJuror, thirdJuror, anyone]) => {
...
+
+              const firstJurorCurrentTotalActiveStake = await registry.jurorsTotalActiveStake(juror)
+              const thirdJurorCurrentTotalActiveStake = await registry.jurorsTotalActiveStake(thirdJuror)
+              assertBn(firstJurorCurrentTotalActiveStake, firstJurorPreviousTotalActiveStake.sub(DRAFT_LOCK_AMOUNT.mul(bn(3))), 'first slashed juror total active balance does not match')
+              assertBn(thirdJurorCurrentTotalActiveStake, thirdJurorPreviousTotalActiveStake.sub(DRAFT_LOCK_AMOUNT.mul(bn(6))), 'second slashed juror total active balance does not match')
+
+              const firstJurorBrightIdCurrentTotalActiveStake = await registry.jurorsTotalActiveStake(jurorBrightIdAddress)
+              const thirdJurorBrightIDCurrentTotalActiveStake = await registry.jurorsTotalActiveStake(thirdJurorBrightIdAddress)
+              assertBn(firstJurorCurrentTotalActiveStake, firstJurorBrightIdCurrentTotalActiveStake, 'first slashed juror total active balance does not match')
+              assertBn(thirdJurorCurrentTotalActiveStake, thirdJurorCurrentTotalActiveStake, 'second slashed juror total active balance does not match')
```
Comment (fabriziovigevani) [inline test/registry/jurors-registry-slashing.js]:
```suggestion
              assertBn(thirdJurorCurrentTotalActiveStake, thirdJurorBrightIDCurrentTotalActiveStake, 'second slashed juror total active balance does not match')
```

File: contracts/registry/JurorsRegistry.sol [hunk 14]
```diff
@@ -14,6 +14,7 @@ import "../standards/ApproveAndCall.sol";
 import "../court/controller/Controller.sol";
 import "../court/controller/ControlledRecoverable.sol";
 
+import "@nomiclabs/buidler/console.sol"; // TODO: Remove
```
Comment (fabriziovigevani) [inline contracts/registry/JurorsRegistry.sol:17]: Should we remove it now?

File: test/registry/jurors-registry-staking.js [hunk 15]
```diff
@@ -146,6 +153,12 @@ contract('JurorsRegistry', ([_, juror, anotherJuror]) => {
 
           itHandlesStakesProperlyFor(amount, data)
         })
+
+        // context('when the juror uses an unverified previous address', () => {
+        //   it('reverts', async () => {
+        //     await assertRevert(registry.stake(MIN_ACTIVE_AMOUNT, data, { from }), 'JR_SENDER_NOT_VERIFIED')
+        //   })
+        // })
```
Comment (fabriziovigevani) [inline test/registry/jurors-registry-staking.js:161]: Should we also remove all this commented tests?

**Review by fabriziovigevani — APPROVED** [review]
A few more comments and LGTM!

File: test/registry/jurors-registry-staking.js (hunk 15 above)
Comment (willjgriff) [reply test/registry/jurors-registry-staking.js:161]: Thanks for pointing it out but nah they're fine for now. I will uncomment some when I add verified account restrictions and will remove the unused ones then.

File: contracts/registry/JurorsRegistry.sol (hunk 14 above)
Comment (willjgriff) [reply contracts/registry/JurorsRegistry.sol:17]: Nah, still using it a bit so will remove in a following PR.

**Pull request merged**